### PR TITLE
Performance work (inlining, scheduler, spinlocks, dlist)

### PR DESCRIPTION
**Pull request by andyross**
Performance work, primarily on the scheduler.

* **First off note that the dlist change in this series technically constitutes a change of API behavior** (though IMHO a good one) so may not be acceptable to everyone.  Please review.  The alternative would be to add a new call or two with better performance characteristics.

* Adds a microbenchmark that focuses specifically on constant factor overhead of scheduler internals.  This test shows about a 35% improvement when the other patches in this series are applied.  And it can be run deterministically using -icount even on current qemu, which is pleasing.

* Remove a few bits from the scheduler preemption test hot path by carefully controlling initial conditions.

* Manually exerts control over the compiler optimizer with ALWAYS_INLINE.  Appently my, heh, exquisitly clean decomposed functions in the new scheduler code were just too much for the compiler to reliably inline, and frankly it was making very bad decisions.  Fixed this up by staring at the generated code and verifying that the CALL sequences were disappearing.

* There was a similar spot in spinlocks, where what should have been a tiny inline sequence became a multilevel function call due to failed inlining.

Honestly the inlining control was really the most important factor here.  I'm not sure whether to spin that as "the code performance was very high already" or "Andy was too dumb to find meaningful optimizations".

### CODE REVIEW

**Review by tyrantlizard77 — APPROVED** [review]
lovely

**Review by dcpleung — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by pabigot — DISMISSED** [review]
I'm only reviewing the dlist change.  In the existing code there are no references to `sys_dlist_insert_after`, and only three references to `sys_dlist_insert_before`, all of which pass a known-valid object for the insertion point.  So I have no objections to removing the check for a null insertion point.

But it's not "technically" an API change--it **is** an API change: the function doesn't do what it used to do, and is no longer documented to do what it used to do.

Make that clear by changing the declaration to remove the now-unused `list` parameter.  Its presence will just confuse people, and changing the arguments will ensure anybody who happens to use this out-of-tree will break at compile time, rather than when the null pointer they passed gets dereferenced.

File: tests/benchmarks/sched/README.rst [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+4. The partner thread then runs and calls _pend_current_thread() again
+   (the kernel switches to the main thread)
+5. The main thread returns from k_yield()
+
+It then iterates this many times, reporting timestamp latencies
+between each numbered step and for the whole cycle, and a running
+average for all cycles run.
+
+Note that because this involves no timer interaction (except, on some
+architectures, k_cycle_get_32()), it works correctly when run in qemu
```
Comment (dbkinder) [inline tests/benchmarks/sched/README.rst]: If I were being picky I'd ask for qemu to be spelled ``QEMU``

File: tests/benchmarks/sched/README.rst [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+using the -icount argument, which can produce 100% deterministic
+behavior (not cycle-exact hardware simulation, but exactly N
+instructions per simulated nanosecond).  You can enable this using an
+environment variable set at cmake time (it's not enough to do this for
```
Comment (dbkinder) [inline tests/benchmarks/sched/README.rst]: I think you're saying this environment variable needs to be set for the build and running of the test, right?  Maybe change this to a positive statement saying this.

**Review by dbkinder — DISMISSED** [review]

File: tests/benchmarks/sched/README.rst (hunk 2 above)
Comment (andyross) [reply tests/benchmarks/sched/README.rst]: No, I'm warning that the environment variable needs to be set for the first stage (cmake) of building the test, not just the second one (make/ninja) where environment variables are commonly interpreted.   Clarified, hopefully.

File: tests/benchmarks/sched/README.rst (hunk 1 above)
Comment (andyross) [reply tests/benchmarks/sched/README.rst]: Fixed

**Comment by andyross** [comment]
Changing the function signature was a bridge too far to my mind.  Zephyr has an API compatibility policy and we're supposed to mark older APIs deprecated for a full release before removing them (and as this coming release is a LTS, even that seems a bit user-hostile).  Also I guess there's a futureproofing argument in that this particular list mechanism doesn't actually need a list pointer to do an insert, some other container styles might in the future (though I can't see why we'd want anything for a dlist other than what we have now).  It was missing an ARG_UNUSED() for sure though, fixed now.

I felt we could cheat around the edges of the rule with the behavior of that weird null edge case, though.

**Comment by pabigot** [comment]
> Changing the function signature was a bridge too far to my mind. Zephyr has an API compatibility policy and we're supposed to mark older APIs deprecated for a full release before removing them (and as this coming release is a LTS, even that seems a bit user-hostile).

Doesn't the API compatibility policy require that the API *behavior* be retained too?

Seriously, keeping the signature the same and changing the implementation so that what used to be a legitimate usage will now dereference a null pointer is pretty user-hostile.  Either add new API that has the behavior you want, or replace the API with something that signals it's different at compile-time.

Personally I support keeping the new behavior and changing the signature; there's plenty of recent precedent for discarding stuff we don't want to retain in LTS.  I've marked this for dev-review on Thursday, so we can get "consensus" there if doing it here isn't enough.

**Comment by andyross** [comment]
(Fix that gitlint issue with the missing colon in the commit message.  Rebased)

**Comment by pabigot** [comment]
Per dev review: @nashif to recommend solution.

**Comment by nashif** [comment]
> Like I said, the alternative is to add another insertion routine for performance reasons, change existing usage, and deprecate the old ones.

lets go with this.

**Comment by andyross** [comment]
OK, updated as requested to make the insert call an entirely new API.  Though I see now that it needs a rebase...

**Comment by nashif** [comment]
timing_info app on e70, first without this patch, second with master:
```
***** Booting Zephyr OS zephyr-v1.13.0-3722-gcb1047351c *****
starting test - Time Measurement
Timing Results: Clock Frequency: 300 MHz
Context switch                               : 154 cycles ,   513 ns
Thread Creation                              : 600 cycles ,  2000 ns
Thread cancel                                : 220 cycles ,   733 ns
Thread abort                                 : 822 cycles ,  2740 ns
Thread Suspend                               : 645 cycles ,  2150 ns
Thread Resume                                : 531 cycles ,  1770 ns
Thread Yield                                 : 512 cycles ,  1706 ns
Thread Sleep                                 :1025 cycles ,  3416 ns
Heap Malloc                                  : 643 cycles ,  2143 ns
Heap Free                                    : 630 cycles ,  2100 ns
Semaphore Take with context switch           :1087 cycles ,  3623 ns
Semaphore Give with context switch           : 916 cycles ,  3053 ns
Semaphore Take without context switch        :  81 cycles ,   270 ns
Semaphore Give without context switch        : 183 cycles ,   610 ns
Mutex lock                                   : 209 cycles ,   696 ns
Mutex unlock                                 : 252 cycles ,   840 ns
Message Queue Put with context switch        : 908 cycles ,  3026 ns
Message Queue Put without context switch     : 306 cycles ,  1020 ns
Message Queue get with context switch        :1052 cycles ,  3506 ns
Message Queue get without context switch     : 294 cycles ,   980 ns
MailBox synchronous put                      :1719 cycles ,  5730 ns
MailBox synchronous get                      :1095 cycles ,  3650 ns
MailBox asynchronous put                     : 479 cycles ,  1596 ns
MailBox get without context switch           : 637 cycles ,  2123 ns
Timing Measurement  finished
PASS - main
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL
***** Booting Zephyr OS zephyr-v1.13.0-3820-gca8d6f21e6 *****
starting test - Time Measurement
Timing Results: Clock Frequency: 300 MHz
Context switch                               : 158 cycles ,   526 ns
Thread Creation                              : 596 cycles ,  1986 ns
Thread cancel                                : 227 cycles ,   756 ns
Thread abort                                 : 783 cycles ,  2610 ns
Thread Suspend                               : 661 cycles ,  2203 ns
Thread Resume                                : 525 cycles ,  1750 ns
Thread Yield                                 : 510 cycles ,  1700 ns
Thread Sleep                                 :1057 cycles ,  3523 ns
Heap Malloc                                  : 642 cycles ,  2140 ns
Heap Free                                    : 631 cycles ,  2103 ns
Semaphore Take with context switch           :1040 cycles ,  3466 ns
Semaphore Give with context switch           : 823 cycles ,  2743 ns
Semaphore Take without context switch        :  82 cycles ,   273 ns
Semaphore Give without context switch        : 163 cycles ,   543 ns
Mutex lock                                   : 207 cycles ,   690 ns
Mutex unlock                                 : 222 cycles ,   740 ns
Message Queue Put with context switch        : 821 cycles ,  2736 ns
Message Queue Put without context switch     : 253 cycles ,   843 ns
Message Queue get with context switch        :1041 cycles ,  3470 ns
Message Queue get without context switch     : 282 cycles ,   940 ns
MailBox synchronous put                      :1650 cycles ,  5500 ns
MailBox synchronous get                      :1007 cycles ,  3356 ns
MailBox asynchronous put                     : 413 cycles ,  1376 ns
MailBox get without context switch           : 577 cycles ,  1923 ns
Timing Measurement  finished
PASS - main
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL

```

**Comment by andyross** [comment]
Heh, I don't claim to make everything faster.  Which commit is which, though?  It wasn't completely clear which one is the patched version and which was master.  I'll have to dig out my SAME70 board.  I did look briefly at esp32 and nrf52840_pca10056 and they were behaving like I expect.

FWIW, the work I was doing was looking at sys_kernel numbers (and also the new microbenchmark), not timing_info (sys_kernel is what the original problem was reported against).  I just re-ran my rig, and these numbers are what I'm seeing there between commits ca8d6f21e6aa (the head of this PR as it stands right now) and addafb8222fb (the master commit on which it was last rebased).

These numbers aren't on hardware, they're from qemu with a QEMU_EXTRA_FLAGS setting of "-icount shift=0,align=off,sleep=off" and a custom rig that feeds qemu's standard input with bytes to keep it running (that works for many of the -icount failures, but not all of them).  With shift=0, each nanosecond represents one CPU instruction:

```
                         arm-base          arm-patched 
     FIFO #1                1050           858 (-18%)
     FIFO #2                 933           783 (-16%)
     FIFO #3                 825           691 (-16%)
     LIFO #1                1050           858 (-18%)
     LIFO #2                 933           783 (-16%)
     LIFO #3                 758           633 (-16%)
Semaphore #1                1008           808 (-20%)
Semaphore #2                 875           725 (-17%)
Semaphore #3                 700           575 (-18%)
    Stack #1                1041           841 (-19%)
    Stack #2                 891           741 (-17%)
    Stack #3                 708           583 (-18%)

                         x86-base          x86-patched 
     FIFO #1                1441          1089 (-24%)
     FIFO #2                1255          1001 (-20%)
     FIFO #3                1089           863 (-21%)
     LIFO #1                1442          1091 (-24%)
     LIFO #2                1257          1002 (-20%)
     LIFO #3                1017           799 (-21%)
Semaphore #1                1348          1001 (-26%)
Semaphore #2                1151           901 (-22%)
Semaphore #3                 912           696 (-24%)
    Stack #1                1386          1039 (-25%)
    Stack #2                1177           927 (-21%)
    Stack #3                 937           721 (-23%)

                      xtensa-base       xtensa-patched 
     FIFO #1                1900          1620 (-15%)
     FIFO #2                1670          1420 (-15%)
     FIFO #3                1540          1330 (-14%)
     LIFO #1                1910          1620 (-15%)
     LIFO #2                1660          1430 (-14%)
     LIFO #3                1420          1220 (-14%)
Semaphore #1                1760          1470 (-16%)
Semaphore #2                1540          1310 (-15%)
Semaphore #3                1300          1100 (-15%)
    Stack #1                1790          1500 (-16%)
    Stack #2                1550          1320 (-15%)
    Stack #3                1300          1100 (-15%)
```

**Comment by andyross** [comment]
While I don't have a SAM e70 handy, I did have a FRDM K64F sitting in a drawer right here, so I repeated the numbers comparison from sys_kernel:

```
                        frdm-base         frdm-patched 
     FIFO #1               23441         19360 (-17%)
     FIFO #2               21125         17943 (-15%)
     FIFO #3               18450         15797 (-14%)
     LIFO #1               23370         19385 (-17%)
     LIFO #2               20925         17793 (-15%)
     LIFO #3               17016         14597 (-14%)
Semaphore #1               21879         17768 (-19%)
Semaphore #2               19175         16022 (-16%)
Semaphore #3               15358         12760 (-17%)
    Stack #1               22908         18714 (-18%)
    Stack #2               19908         16635 (-16%)
    Stack #3               15762         13210 (-16%)
```

And did a quick one for timing_info:


```
Context switch                              260 230 -11%
Heap Free                                   919 899 -01%
Heap Malloc                                 980 933 -04%
MailBox asynchronous put                    530 440 -16%
MailBox get without context switch          920 695 -23%
Message Queue Put without context switch    375 300 -19%
Message Queue get without context switch    370 280 -23%
Mutex lock                                  340 275 -18%
Mutex unlock                                420 321 -23%
Semaphore Give without context switch       280 215 -22%
```

Again, these are between the pull request HEAD and the master commit directly under it.  These numbers look pretty much exactly like what I'd expect.  Not sure what you're seing with the Atmel board, maybe there's a confounding commit merged somewhere that's slowed things down?

**Comment by rljordan** [comment]
@andyross  Can this be closed?

**Pull request merged**